umami: init at 2.19.0; nixos/umami: init

diogotcorreia opened the pull request:
Add Umami (closes #172063)

I've taken the same approach as @jlesquembre, but improved on it to run migrations automatically when Umami is started. I've also solved the problem with upstream making HTTP requests to get GeoCities during the build process by pinning it to a specific revision when updating (see update script).

As for some pesky details:
- Umami needs the database variant to be defined at build time since it uses different schema definitions for postgres and mysql, so I make it an argument to the derivation, which is override-able from a future module.
- Likewise, the environment variables `COLLECT_API_ENDPOINT` and `TRACKER_SCRIPT_NAME` pose a challenge as they are read during build time. Upstream's docker image goes around this by adding a [middleware file](https://github.com/umami-software/umami/blob/v2.15.1/docker/middleware.js) for this to happen at runtime, but I decided against that since it would (slightly) reduce performance, it doesn't take that long to rebuild umami (~2 min on my laptop), and it probably is a niche feature. Additionally, if done at runtime, setting `COLLECT_API_ENDPOINT` would require changing a file in the derivation, so we would have to perform some hacks to do that.
- While I've tested the PostgreSQL variant successfully, ~~I could not test MySQL~~ since I used MariaDB and there seems to be an [upstream issue with one of the migrations](https://github.com/umami-software/umami/issues/2645). ~~I might try again in the next few days.~~ **EDIT:** MySQL works fine; MariaDB doesn't, but that's a problem upstream.

Example to override build arguments:
```nix
pkgs.umami.override {
  databaseType = "mysql";
  collectApiEndpoint = "/api/alternate-send",
  trackerScriptNames = [ "tracker.js" "tracker2.js" ],
}
```

I've tested this with a clean database (PostgreSQL) and **it seems to work perfectly**. I will try it in my production deployment in the next few days and perhaps do a module as well. **EDIT:** Has been in production for a while, working fine.

Previous attempts (pinging relevant people so they can chime in):
Closes #371889 @elikoga @GiggleSquid
Closes #372220 @jlesquembre

Feedback/testing is very much appreciated. In theory, the only thing necessary to run this is to set `DATABASE_URL` and then run `result/bin/umami-server`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

diogotcorreia committed:
umami: init at 2.19.0

diogotcorreia commented:
Update: I've tried rebasing this over the latest commits on master and it seems like it broke due to a Prisma version mismatch (I was testing with 6.0.1 and now it's at 6.3.0). Umami is still at 5.22.0 in the latest version, but the dev branch is already targeting 6.1.0, which hopefully is compatible.

I'll investigate further tomorrow, so I won't push the rebase now.

GiggleSquid commented:
This looks promising. I'll do some basic funcion testing over the next few days, and report back. 

Great job, very excited!

diogotcorreia commented:
I've manage to fix the breakage when rebasing over master by pinning `prisma-engines`. Should be ready :)

I have also finished a NixOS module that I'll push in a sec after I finalize some testing.

diogotcorreia committed:
nixos/umami: init module

diogotcorreia committed:
doc: add services.umami to 25.11 release notes

diogotcorreia commented:
I've added the module, reviews and other commends are appreciated. It is a bit opinionated when it comes to handling secrets, let me know if anything should be changed to better fit in nixpkgs.
Hardening might also need some improvement, but the score is pretty good already:
```
Overall exposure level for umami.service: 3.0 OK 🙂
```

I'll be adding tests tomorrow and then the PR should be complete.

PS: just [deployed this to production](https://github.com/diogotcorreia/dotfiles/commit/8ce5f2bf7b14dfa5a82dccabb3849abee135851b), working perfectly! :)

tecosaur commented:
I have a WIP Umami module that may be of some interest:

<details>
<summary>
module.nix
</summary>

```nix
{ config, lib, pkgs, ... }:

# See:
# - <https://umami.is>
# - <https://github.com/jonaharagon/caddy-umami>

let
  umami-pkg = pkgs.callPackage ../packages/umami.nix { };
  cfg = config.services.umami;
  createDb = cfg.database.createLocally && cfg.database.host == null;
  postgresqlPkg = if config.services.postgresql.enable then config.services.postgresql.package else pkgs.postgresql;
in {
  options.services.umami = {
    enable = lib.mkEnableOption "Umami analytics";
    package = lib.mkOption {
      type = pkgs.lib.types.package;
      default = umami-pkg;
      description = "The umami package to use";
    };
    user = lib.mkOption {
      type = lib.types.str;
      default = "umami";
      description = lib.mdDoc "The user to run the umami service as";
    };
    group = lib.mkOption {
      type = lib.types.str;
      default = "umami";
      description = lib.mdDoc "Primary group of the umami user";
    };
    database = {
      name = lib.mkOption {
        type = lib.types.str;
        default = "umami";
        description = lib.mdDoc "The name of the umami database";
      };
      host = lib.mkOption {
        type = with lib.types; nullOr str;
        default = null;
        description = ''
          Umami database hostname. `null` means
          “prefer local unix socket connection”.
        '';
      };
      createLocally = lib.mkOption {
        type = lib.types.bool;
        default = true;
        description = lib.mdDoc "Whether to create the umami database";
      };
    };
  };

  config = lib.mkIf cfg.enable {
    systemd.services.umami = {
      description = "Umami analytics";
      after = [ "network-online.target" "postgresql.service" "umami-postgres-setup.service" ];
      bindsTo =
        lib.optionals (cfg.database.host == null) [
          "postgresql.service"
          "umami-postgres-setup.service"
        ];
      wants = [ "network-online.target" ];
      wantedBy = [ "multi-user.target" ];
      environment = {
        PORT = toString config.site.apps.umami.port;
        DATABASE_URL = "postgresql://${cfg.user}@localhost/umami?host=/run/postgresql";
        PRISMA_SCHEMA_ENGINE_BINARY = "${pkgs.prisma-engines}/bin/schema-engine";
        PRISMA_QUERY_ENGINE_BINARY = "${pkgs.prisma-engines}/bin/query-engine";
        PRISMA_QUERY_ENGINE_LIBRARY = "${pkgs.prisma-engines}/lib/libquery_engine.node";
        PRISMA_INTROSPECTION_ENGINE_BINARY = "${pkgs.prisma-engines}/bin/introspection-engine";
        PRISMA_FMT_BINARY = "${pkgs.prisma-engines}/bin/prisma-fmt";
      };
      path = with pkgs; [
        cfg.package
        openssl
        prisma-engines
        postgresqlPkg
      ];
      serviceConfig = {
        DynamicUser = true;
        User = cfg.user;
        ExecStartPre = [
          "${cfg.package}/node_modules/.bin/prisma migrate deploy"
          "${cfg.package}/node_modules/.bin/prisma db seed"
        ];
        ExecStart = "${cfg.package}/node_modules/.bin/next start -p $PORT";
        StateDirectory = "umami";
        StandardOutput = "journal";
      };
    };

    users.users = lib.optionalAttrs (cfg.user == "umami") {
      umami = {
        isSystemUser = true;
        group = cfg.group;
      };
    };

    users.groups = lib.optionalAttrs (cfg.group == "umami") {
      umami = { };
    };

    services.postgresql = lib.mkIf createDb {
      enable = true;
      ensureUsers = [ { name = cfg.user; } ];
    };

    systemd.services.umami-postgres-setup = lib.mkIf createDb {
      description = "Umami analytics PostgreSQL setup";
      after = [ "postgresql.service" ];
      requires = [ "postgresql.service" ];
      bindsTo = [ "postgresql.service" ];
      wantedBy = [ "umami.service" ];
      partOf = [ "umami.service" ];
      path = [ postgresqlPkg ];
      script = ''
        set -o errexit -o pipefail -o nounset -o errtrace
        shopt -s inherit_errexit

        psql -tAc "SELECT 1 FROM pg_database WHERE datname = '${cfg.database.name}'" | grep -q 1 || psql -tAc 'CREATE DATABASE "${cfg.database.name}" OWNER "${cfg.user}"'
        psql '${cfg.database.name}' -tAc "CREATE EXTENSION IF NOT EXISTS pg_trgm"
        psql '${cfg.database.name}' -tAc "CREATE EXTENSION IF NOT EXISTS hstore"
      '';
      serviceConfig = {
        User = config.services.postgresql.superUser;
        Type = "oneshot";
        RemainAfterExit = true;
      };
    };
  };
}
```

</details>

diogotcorreia commented:
Thanks @tecosaur, that helps! There are a few noticeable differences:
- You use the scripts in `node_modules/.bin`, but that doesn't exist in the way we build the package now. Additionally, we run the migrations in the package and not in the module like you are doing :sweat_smile: In my opinion, it makes more sense for the package to work on its own, but let me know if you have a different view;
- You allow specifying the user/group for umami, but it doesn't need any state so I'm just using `DynamicUser` (which still allows for peer auth to the database). Would it make sense to allow a user to be set? I think it would just be adding complexity unnecessarily;
- Same thing for the database name, I'm just assuming it's hardcoded and will let the user handle it themselves (through `DATABASE_URL`) if they want something else;
- Speaking of the database, your module setups the `pg_trgm` and `hstore` extensions, but I can't find this in Umami docs. Are these really needed?
- As for the prisma engine env vars, I tested and it seems only `PRISMA_QUERY_ENGINE_LIBRARY` is needed, so that's why I only set that (in the package). The other variables seem to not be used in production at least;
- As for seeding, upstream doesn't seem to do it, so I left it out;
- You have `after = [ "postgresql.service" ];` which is probably good to add, even though it's working fine for me without.

Let me know if you have a strong opinion on any of these differences, and I'll update my module accordingly :)

PS: I noticed umami supports redis for caching, I can probably add an option to set that up automatically as well.

diogotcorreia committed:
nixos/tests/umami: init

diogotcorreia commented:
One last change, I managed to use the prisma cli from node_modules by patching the next config to include it in the final build. This avoids version mismatches with the one in nixpkgs.
Unfortunately we still need to override `prisma-engines`, but there's no way around that one. And I wanted to get the update script to handle pinning of `prisma-engines`, but there doesn't seem to exist a `prefetch-cargo-deps`, so that'll have to be done manually for now :pensive:

aqrln commented:
> As for the prisma engine env vars, I tested and it seems only PRISMA_QUERY_ENGINE_LIBRARY is needed, so that's why I only set that (in the package). The other variables seem to not be used in production at least;

That's correct. You also need `PRISMA_SCHEMA_ENGINE_BINARY` to run `prisma migrate deploy` but only `PRISMA_QUERY_ENGINE_LIBRARY` is used in the application yourself at run time.

`PRISMA_QUERY_ENGINE_BINARY` exists but you don't need it, and the other two don't even exist and will not be checked by Prisma.

diogotcorreia commented:
I've updated this to Umami 2.16 (will squash into the first commit upon approving review)
https://github.com/umami-software/umami/releases/tag/v2.16

diogotcorreia commented:
https://github.com/umami-software/umami/releases/tag/v2.16.1
2.16.1 was released, updated as well

diogotcorreia commented:
I've updated Umami to [v2.17.0](https://github.com/umami-software/umami/releases/tag/v2.17.0) and fixed a merge conflict by rebasing over master. I've also squashed all the fixup commits, since there are no in-progress reviews.

I'd appreciate that someone interested in getting this merged would help with a review :)

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 380249`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>umami</li>
  </ul>
</details>

elikoga commented:
I haven't gotten the chance to run this umami yet, however the source looks good. The [update.sh](https://github.com/NixOS/nixpkgs/pull/380249/files#diff-de0155cefc424592ac24697338cf19fcf205113f6e05a1b4d7fddc2714350b56) is a nice touch. I am not quite sure wether `cc-by-40 # geocities` is correct as the license for maxminds geo db

74k1 commented:
Been using this Module for a few days now and it's been working great! :+1: LGTM

diogotcorreia commented:
@elikoga 
> I am not quite sure wether `cc-by-40 # geocities` is correct as the license for maxminds geo db

IANAL, but the README of https://github.com/GitSquared/node-geolite2-redist says:
> The databases themselves are provided by MaxMind under [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/)
> 
> For the library, see [LICENSE](https://github.com/GitSquared/node-geolite2-redist/blob/master/LICENSE).

They, however, also list limitations and have a full EULA: https://www.maxmind.com/en/geolite2/eula
Point 3 states:
> Subject to the terms and conditions of this Agreement, to the extent the Services contain any copyrightable elements those copyrightable elements are governed by the Creative Commons License.

Not sure what to make of this, if someone is more familiar with this I'd appreciate the help

axelkar commented on pkgs/by-name/um/umami/package.nix:
```diff
@@ -0,0 +1,180 @@
+{
+  lib,
+  stdenv,
```
Hi. Shouldn't `stdenvNoCC` be used instead?

diogotcorreia replied:
Yep, that works fine, thanks for noticing!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/packaging-umami-analytics-when-build-requires-postgres-prisma-orm/62789/2

diogotcorreia commented:
Fixed merge conflicts (rebased over master)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5369

diogotcorreia commented:
https://github.com/umami-software/umami/releases/tag/v2.18.0

Upgraded to Umami 2.18.0.
Also, changed the upgrade script so that it can handle updating the prisma-engines pin, meaning it should be fully unattended now!

Umami has changed from yarn to pnpm with this release, and so I had a lot of problems getting the `node_modules/.bin/prisma` script to work. I settled on a hack where I copied stuff manually from the node_modules dir, but I really don't like it.
Another solution would be to rely on `prisma` that is packaged in nixpkgs, but we'd need to pin it as well. I'm unsure if it is easy to override `pnpmDeps.hash`, so that could be a problem. Another problem with that route is that the derivation might change to the point that simply pinning the `src` and `pnpmDeps.hash` doesn't work anymore.
Either way, I'm open to suggestions.

chpatrick commented:
Could you also add runtime setting of BASE_PATH?

diogotcorreia commented:
@chpatrick Done!

diogotcorreia commented:
Done. Also fixed merge conflicts and moved the entry from the 25.05 release notes to 25.11.

ners commented on pkgs/by-name/um/umami/package.nix:
```diff
@@ -0,0 +1,200 @@
+  # install dev dependencies as well, for rollup
+  pnpmInstallFlags = [ "--prod=false" ];
+
+  pnpmDeps = pnpm.fetchDeps {
+    inherit (finalAttrs)
+      pname
+      pnpmInstallFlags
+      version
+      src
+      ;
+    hash = "sha256-WkSMA18QapbpYe2FMabD4yNmbg0WRdnymZHfv1VOjSk=";
+  };
```
```suggestion
  pnpmDeps = pnpm.fetchDeps {
    inherit (finalAttrs)
      pname
      pnpmInstallFlags
      version
      src
      ;
    fetcherVersion = 2;
    hash = "sha256-WkSMA18QapbpYe2FMabD4yNmbg0WRdnymZHfv1VOjSk=";
  };
```

ners replied:
Build fails on latest unstable without this.

karaolidis commented:
Can someone confirm whether the DISABLE_LOGIN environment variable works for them? I am testing it with the following container:

```nix
{ pkgs, ... }:
let
  # FIXME: https://github.com/NixOS/nixpkgs/pull/380249/files/453c3be35adc9161a4e1c9247a22a6f4bfdd6fb1#r2215773111
  umami = pkgs.umami.overrideAttrs (oldAttrs: {
    pnpmDeps = pkgs.pnpm.fetchDeps {
      inherit (oldAttrs)
        pname
        pnpmInstallFlags
        version
        src
        ;
      fetcherVersion = 2;
      hash = "sha256-GXDnLBfPN7eC1bg+dsCHf38U7oWt1P+MGWZyV/Hn3QM=";
    };
  });
in
pkgs.dockerTools.buildImage {
  name = "umami";
  fromImage = import ../base { inherit pkgs; };

  copyToRoot = pkgs.buildEnv {
    name = "root";
    paths = [ umami ];
    pathsToLink = [ "/bin" ];
  };

  config = {
    Entrypoint = [ "umami-server" ];
    Volumes = {
      "/var/lib/umami" = { };
    };
    WorkingDir = "/var/lib/umami";
    ExposedPorts = {
      "3000/tcp" = { };
    };
  };
}
```

```sh
nix build .#docker-umami
podman run --rm -it -p 3000:3000 --network umami -e DISABLE_LOGIN=true -e DATABASE_URL=postgresql://umami:umami@postgres:5432/umami docker-archive:./result
```

nh2 commented on pkgs/by-name/um/umami/package.nix:
```diff
@@ -0,0 +1,201 @@
+    cp -a node_modules/.pnpm/prisma* $out/node_modules/.pnpm
+    # remove broken symlinks (some dependencies that are not relevant for running migrations)
+    find $out/node_modules/.pnpm/@prisma* -xtype l -exec rm {} \;
+    find $out/node_modules/.pnpm/prisma* -xtype l -exec rm {} \;
```
Theoretically for full shell correctness, all variables like `$out" should be quoted.

We currently know that `$out` in `/nix/store` generally does not contain spaces, but in the future people may want to use nix with custom prefixes that contain spaces.

I'll push a commit that quotes at least the two recursive `rm`'s above, because those could be truly disastrous if used with spaces in `$out`.

nh2 reviewed (approved):
Looks good and tests pass. Merging (I enabled auto-merge after checks pass, so it should do it soon).

Thank you very much!
